### PR TITLE
Add Transliteration feature in tts-middleware

### PR DESCRIPTION
**Pull request by iamsh4shank**
In TTS middleware, add the functionality of transliteration for Hindi VTTS.

Here we are using the [Indic-trans](https://github.com/indic-transliteration/indic_transliteration_py) library that generated the proper utterance for Hindi to English

### CODE REVIEW

File: tts_middleware/core.py [hunk 1]
```diff
@@ -21,8 +38,13 @@ def tts_middleware(tts_function):
     def _tts(text: str, language_code: str) -> Audio:
         node = pq(text)
         raw_text = node.text()
+
+        lang = supported_langs[language_code]
+
+        trans = Transliterator(source=lang, target='eng', build_lookup=True)
```
Comment (prashantskit) [inline tts_middleware/core.py:44]: Shouldn't the target language be "hin"

Comment (iamsh4shank) [reply tts_middleware/core.py:44]: ok, I got it, I will modify this

File: tts_middleware/core.py [hunk 2]
```diff
@@ -12,6 +16,25 @@
...
+    reg = re.compile(r'[a-zA-Z]')
+    trans = Transliterator(source='eng', target='hin', build_lookup=True)
+    trans_text = ""
+    lang_code = {}
+    for word in text.split(" "):
+        if word not in '[@_!#$%^&*()<>?/\|}{~:]' and reg.match(word):
+            lang_code[word] = "en"
+        else: 
+            lang_code[word] = "hi"
+    for key in lang_code:
```
Comment (pskrunner14) [inline tts_middleware/core.py:29]: Can we use filter, map, and join functions to make this code cleaner?

File: tts_middleware/core.py [hunk 3]
```diff
@@ -12,6 +16,25 @@
 Audio = Tuple[np.ndarray, int]
 
 
+def transliteration(text):
+    reg = re.compile(r'[a-zA-Z]')
+    trans = Transliterator(source='eng', target='hin', build_lookup=True)
```
Comment (pskrunner14) [inline tts_middleware/core.py:21]: You're assuming the target language is always `hi`. Let's use the `language_code` parameter and only transliterate when it actually is `hi`.

File: tts_middleware/core.py [hunk 4]
```diff
@@ -21,8 +44,11 @@ def tts_middleware(tts_function):
     def _tts(text: str, language_code: str) -> Audio:
```
Comment (pskrunner14) [inline tts_middleware/core.py:44]: Maybe add a flag parameter for toggling transliteration? We don't always want to trigger transliteration

File: tts_middleware/core.py [hunk 5]
```diff
@@ -12,6 +16,25 @@
...
+    lang_code = {}
+    for word in text.split(" "):
+        if word not in '[@_!#$%^&*()<>?/\|}{~:]' and reg.match(word):
+            lang_code[word] = "en"
+        else: 
+            lang_code[word] = "hi"
+    for key in lang_code:
+        #print (keys)
+        if lang_code[key] == "en":
+            trans_text += f'{trans.transform(key)} '
```
Comment (pskrunner14) [inline tts_middleware/core.py:32]: Let's also test the transform function for performance. If it turns out to be slower than expected, we should think of optimizations like caching and parallelizing the call.

**Review by pskrunner14 — CHANGES_REQUESTED** [review]

File: tts_middleware/core.py (hunk 5 above)
Comment (iamsh4shank) [reply tts_middleware/core.py:32]: Sure as of now I tested for a few utterance from the sheet and performed quite fast with them

File: tts_middleware/core.py (hunk 4 above)
Comment (iamsh4shank) [reply tts_middleware/core.py:44]: Ok sure

**Comment by iamsh4shank** [comment]
@pskrunner14 did the following changes can you review it once, thank you :)

File: tts_middleware/core.py [hunk 6]
```diff
@@ -12,15 +14,30 @@
...
     """
     Decorator function for allowing SSML tags for a TTS.
     """
 
     @wraps(tts_function)
-    def _tts(text: str, language_code: str) -> Audio:
+    def _tts(text: str, language_code: str, transliterate: bool) -> Audio:
         node = pq(text)
         raw_text = node.text()
+        if transliterate and language_code == 'hin':
```
Comment (pskrunner14) [inline tts_middleware/core.py:38]: Language code check should be for `hi`, maybe create a mapping that maps this to `hin`

File: tts_middleware/core.py [hunk 7]
```diff
@@ -12,15 +14,30 @@
 Audio = Tuple[np.ndarray, int]
 
 
+def transliteration(text, lang_code):
+    reg = re.compile(r'[a-zA-Z]')
+    trans = Transliterator(source='eng', target=lang_code, build_lookup=True)
+    trans_text = ""
+    for word in text.split(" "):
```
Comment (pskrunner14) [inline tts_middleware/core.py:21]: For loop will be really slow for our use case. Let's move to map and join

**Review by pskrunner14 — CHANGES_REQUESTED** [review]

**Comment by pskrunner14** [comment]
We're also moving to google transliteration API I'm assuming? @iamsh4shank

**Comment by iamsh4shank** [comment]
@pskrunner14 modified the code with the new transliteration API

**Review by pskrunner14 — CHANGES_REQUESTED** [review]

File: tts_middleware/core.py [hunk 8]
```diff
@@ -12,15 +16,28 @@
 Audio = Tuple[np.ndarray, int]
 
 
+def eng_to_hin(word, lang_code):
+    reg = re.compile(r'[a-zA-Z]')
+    if word not in '[@_!#$%^&*()<>?/\|}{~:]' and reg.match(word):
+        return transliterate_text(word, lang_code)
```
[the remark this answers is not in the document]
Comment (iamsh4shank) [reply tts_middleware/core.py:22]: * Processing time is around 0.37 seconds 
* There is no such direct limit that is mentioned.

**Comment by iamsh4shank** [comment]
@pskrunner14 Done, let me know if there are any other changes.

File: tts_middleware/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (prashantskit) [reply tts_middleware/core.py]: the processing time is after parallelising with threads?

File: tts_middleware/text.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+from google.transliteration import transliterate_text
+
+def eng_to_hin(word, lang_code):
+    reg = re.compile(r'[a-zA-Z]')
+    if reg.match(word):
+        return transliterate_text(word, lang_code)
+    return word
+
+def transliteration(text, lang_code):
+    with ThreadPoolExecutor(max_workers=8) as executor:
```
Comment (prashantskit) [inline tts_middleware/text.py:13]: remove the hardcoded value for workers. Do a check on the cores available and limit it to 8.

**Comment by pskrunner14** [comment]
Have made a pre-release with this feature, will integration test it with tts-service later today and do a stable release.

**Comment by pskrunner14** [comment]
@prashantskit one thing I'm concerned about here is the total pipeline latency, this will be a considerable hit to our sub ~600ms goal

**Comment by prashantskit** [comment]
> 

It does but we can implement cache and single API call to fetch the transliteration to speed it up.

**Comment by pskrunner14** [comment]
@iamsh4shank in the future, please create branches on this same org repo and create MR instead of forking and creating PR.

**Comment by iamsh4shank** [comment]
Sure will take care of this

**Pull request merged**